src/main/java/org/roaringbitmap/RoaringIterator.java

```diff
@@ -0,0 +1,138 @@
+package org.roaringbitmap;
+
+import java.util.NoSuchElementException;
+
+class RoaringIterator implements IntIterator {
+  final private RoaringBitmap parent;
+  private ArrayContainer ac;
+  private RunContainer rc;
+  private int containerIndex;
+  private short typecode = TypeCode.NONE;
+  private int hb;
+  private int le = 0;
+  private int maxlength;
+  private int base;
+  private long w;
+  private int pos;
+  private long[] bitmap;
+
+  RoaringIterator(RoaringBitmap parent) {
+    this.parent = parent;
+    this.hb = 0;
+    this.containerIndex = 0;
+    nextContainer();
+  }
+
+  public boolean hasNext() {
+    return containerIndex <= parent.highLowContainer.size;
+  }
+
+  public int next() {
+    int answer;
+    switch (typecode) {
+      case TypeCode.BITSET_CONTAINER:
+        long t = w & -w;
```

Comment:
@openkappa I think that @ppiotrow  explains it well. 

The current iterators have polymorphic calls and lots of short-lived temporary objects. Each and every call to ``next`` goes through a lot of code, not to mention the pressure it puts on the garbage collector. We can hope for the JVM to optimize it all out, but I am pessimistic.


It is not always inefficient,  but in some cases, I suspect that it is.

Would replacement code be less elegant? Possibly, but it is not like we need to play with this code a lot. I think we mostly want to it to be fast.